doc/check_signature.1

CHECK_SIGNATURE (1)                 ROA-PKI                   CHECK_SIGNATURE(1)


NAME

        check_signature - checks the signature on a certificate or a CRL

SYNOPSIS

        check_signature signedfile signerfile

DESCRIPTION

        The signature in the "signedfile" is checked using the public key in
        the "signerfile".  The type of contents in the "signedfile" are
        designated by its suffix: ".cer indicates a certificate, and ".crl"
        indicates a CRL.

        Response messages:

          Usage: signed cert or CRL, signer's cert
          Signature succeeded
          Signature failed
          Can't get file ...
          Signature checking error in ...
          Invalid type ...
          Invalid algorithm ...

       where the ellipsis is replaced with specific information.


ChangeLog
0.10

	* Add beta support for running rpstir in 64-bit mode.
	* Add beta support for running rpstir on Mac OS X.
	* Significantly improve performance of incremental updates by adding an
	  rsync flag to preserve file modification times.
	* Fix multiple potential SQL-injection bugs.
	* Support newer versions of rsync and automake.
	* Remove support for RPSL due to lack of demand.
	* Update conformance cases (see doc/conformance-cases).
	* The ./configure script no longer prints a misleading error message
	  when it's run outside of a git directory.
	* Add tests with AS numbers that don't fit into 16 bits.
	* Reject trust anchor CA certificates that have resources marked
	  inherit, as specified in RFC 6490, Section 2.2.
	* Verify that rpstir compiles with clang, and fix all of clang's
	  warnings.
	* In the statistics collection mode, add support for incremental updates
	  and for running multiple simultaneous statistics collections.
	* Change handling of CRLs that list syntactically invalid serial
	  numbers. These CRLs are now accepted, but the certificates with
	  syntactically invalid serial numbers are still not accepted.

0.9

	* Add many more conformance test cases. (See doc/conformance-cases for
	  the full list.) These test cases can be used by relying party
	  software to test compliance with published RFCs and Internet-Drafts.
	  Additionally, relying party software that passes the tests can be
	  used to test the output of a Certification Authority.
	* Fix bugs found by the above test cases.
	* Add a pseudo-random factor to the calculation of how long to wait
	  before retrying an rsync connection. This should help prevent many
	  relying parties from hitting the same server at the same time.
	* Implement basic support for collecting statistics of the RPKI over
	  time.

0.8

	* Support Ghostbusters Records [RFC 6493].
	* Fix a SQL injection bug when a filename in a manifest contains certain
	  characters.
	* Fix a bug where CRLs were not being marked as invalid when their
	  issuers became invalid and manifests were not being marked as invalid
	  when their associated EE certificates became invalid.
	* Fix the query client (rpstir-query) to not consider it an error
	  when no results are returned for a query.
	* Require ROA EE certificates to explicitly specify IP resources.
	  That is, disallow RFC 3779 "inherit" bit in ROA EE certificates.
	* Add functionality to rpstir-results for outputting simple lists
	  of good and bad URIs, to simplify comparison testing.
	* On installation, create ${bindir} if it does not already exist.
	* Improve README based on user feedback.

0.7

	* Add preliminary support for adding files from the local cache. In the
	  future, this could be used to quickly deploy additional relying party
	  machines without each one needing to download all files from the
	  global RPKI.
	* Fix a buffer overflow when manifests list filenames longer than 199
	  bytes.
	* Fix a read of uninitialized memory when the first file listed on a
	  manifest has an unknown extension.
	* Don't allow empty sets of IP resources in certificates.
	  (http://tools.ietf.org/html/rfc6487#section-4.8.10)
	* Don't allow Subsequent Address Family Identifier (SAFI) values in IP
	  resources. (http://tools.ietf.org/html/rfc6487#section-4.8.10)
	* Fix bugs in the ROA validation code where AS numbers equal to zero
	  or larger than 2^31 - 1 were rejected.

0.6

	* Simplify the commands used during normal operation. The commands
	  are now named after what the user intends to do, instead of after the
	  internal component we used to implement them. Some of the scripts are
	  reorganized so that each script does one function (e.g., synchronize
	  with remote repositories) instead of presenting an interface to one
	  internal component (e.g., the component that validates objects fed to
	  it by other components while syncronizing with remote repositories).
	* Use a single configuration file syntax instead of a mix of
	  environment variables and multiple incompatible configuration files.
	  Additionally, provide an example configuration file with comments
	  describing each option. This should make rpstir easier to set up.
	* Switch from one Makefile per directory to a single Makefile. Aside
	  from speeding up parallel compilation, this lets us simplify
	  internal dependencies. With one Makefile per directory, we had to
	  carefully order the directories such that nothing in one directory
	  depended on something in a later directory. Since the dependencies
	  changed over time, this became more and more complex to maintain.
	  With a single Makefile, we can just tell "make" about each dependency
	  and let "make" figure out the order to build things.
	* Remove duplicate code. Before switching to a single Makefile (see
	  above), we occasionally copied functions from one file to another to
	  avoid having to restructure the source tree and reorder the
	  per-directory Makefiles. With a single Makefile keeping track of
	  dependencies, we consolidated the duplicated code and let "make"
	  do the build ordering for us.
	* Find and remove unused code.
	* Improve support for CHECKTOOL. When $CHECKTOOL is set to "valgrind"
	  for the self-tests, much more code is now tested with valgrind.
	* Make it possible to install to the system instead of running from
	  within the source directory.
	* When an rsync download of a trust anchor fails, retry a few times
	  instead of giving up on downloading that trust anchor.
	* Support Cryptlib 3.4.2.
	* Fix a bug that occured when a trust anchor was added after its
	  children. Some of the children that should have been considered valid
	  were considered unknown. This doesn't happen in normal operation, but
	  was discovered during testing at IETF 85.
	* Don't require access to the database root account. As long as
	  rpstir's database user has full permissions on a database, there's
	  no reason to require access to the root database user.

0.5


cg/tools/check_signature.c
/*
 * $Id: check_signature.c c 506 2008-06-03 21:20:05Z gardiner $ 
 */


#include "roa_utils.h"
#include "crlv2.h"
#include "cryptlib.h"
#include "blob.h"

char *msgs[] = {
    "Signature %s\n",
    "Usage: signed cert or CRL, signer's cert\n",
    "Can't get file %s\n",
    "Signature checking error in %s\n",
    "Invalid type %s\n",
    "Invalid algorithm %s\n",
};

static void fatal(
    int err,
    char *param)
{
    fprintf(stderr, msgs[err], param);
    if (err)
        exit(err);
}

static int gen_hash(
    uchar * inbufp,
    int bsize,
    uchar * outbufp,
    CRYPT_ALGO_TYPE alg)
{                               // used for manifests alg = 1 for SHA-1; alg = 
                                // 2 for SHA2
    CRYPT_CONTEXT hashContext;
    uchar hash[40];
    int ansr = -1;

    if (alg != CRYPT_ALGO_SHA && alg != CRYPT_ALGO_SHA2)
        fatal(3, "hashing");


    memset(hash, 0, 40);
    if (cryptInit() != CRYPT_OK)
        fatal(2, "CryptInit");;
    cryptCreateContext(&hashContext, CRYPT_UNUSED, alg);
    cryptEncrypt(hashContext, inbufp, bsize);
    cryptEncrypt(hashContext, inbufp, 0);
    cryptGetAttributeString(hashContext, CRYPT_CTXINFO_HASHVALUE, hash, &ansr);
    cryptDestroyContext(hashContext);
    cryptEnd();
    memcpy(outbufp, hash, ansr);
    return ansr;
}

static int check_signature(
    struct casn *locertp,
    struct Certificate *hicertp,
    struct casn *sigp)
{
    CRYPT_CONTEXT pubkeyContext,
        hashContext;
    CRYPT_PKCINFO_RSA rsakey;
    // CRYPT_KEYSET cryptKeyset;
    struct RSAPubKey rsapubkey;
    int bsize,
        ret,
        sidsize;
    uchar *c,
       *buf,
        hash[40],
        sid[40];

    // get SID and generate the sha-1 hash
    // (needed for cryptlib; see below)
    memset(sid, 0, 40);
    bsize = size_casn(&hicertp->toBeSigned.subjectPublicKeyInfo.self);
    if (bsize < 0)
        fatal(3, "lo cert size");
    buf = (uchar *) calloc(1, bsize);
    encode_casn(&hicertp->toBeSigned.subjectPublicKeyInfo.self, buf);
    sidsize = gen_hash(buf, bsize, sid, CRYPT_ALGO_SHA);
    free(buf);

    // generate the sha256 hash of the signed attributes. We don't call
    // gen_hash because we need the hashContext for later use (below).
    memset(hash, 0, 40);
    bsize = size_casn(locertp);
    if (bsize < 0)
        fatal(3, "sizing toBeSigned");
    buf = (uchar *) calloc(1, bsize);
    encode_casn(locertp, buf);

    // (re)init the crypt library
    if (cryptInit() != CRYPT_OK)
        fatal(2, "CryptInit");;
    cryptCreateContext(&hashContext, CRYPT_UNUSED, CRYPT_ALGO_SHA2);
    cryptEncrypt(hashContext, buf, bsize);
    cryptEncrypt(hashContext, buf, 0);
    cryptGetAttributeString(hashContext, CRYPT_CTXINFO_HASHVALUE, hash, &ret);
    free(buf);

    // get the public key from the certificate and decode it into an RSAPubKey
    readvsize_casn(&hicertp->toBeSigned.subjectPublicKeyInfo.subjectPublicKey,
                   &c);
    RSAPubKey(&rsapubkey, 0);
    decode_casn(&rsapubkey.self, &c[1]);        // skip 1st byte (tag?) in BIT 
                                                // STRING
    free(c);

    // set up the key by reading the modulus and exponent
    cryptCreateContext(&pubkeyContext, CRYPT_UNUSED, CRYPT_ALGO_RSA);
    cryptSetAttributeString(pubkeyContext, CRYPT_CTXINFO_LABEL, "label", 5);
    cryptInitComponents(&rsakey, CRYPT_KEYTYPE_PUBLIC);

    // read the modulus from rsapubkey
    bsize = readvsize_casn(&rsapubkey.modulus, &buf);
    c = buf;
    // if the first byte is a zero, skip it
    if (!*buf)
    {
        c++;
        bsize--;
    }
    cryptSetComponent((&rsakey)->n, c, bsize * 8);
    free(buf);

    // read the exponent from the rsapubkey
    bsize = readvsize_casn(&rsapubkey.exponent, &buf);
    cryptSetComponent((&rsakey)->e, buf, bsize * 8);
    free(buf);

    // set the modulus and exponent on the key
    cryptSetAttributeString(pubkeyContext, CRYPT_CTXINFO_KEY_COMPONENTS,
                            &rsakey, sizeof(CRYPT_PKCINFO_RSA));
    // all done with this now, free the storage
    cryptDestroyComponents(&rsakey);

    // make the structure cryptlib likes.
    // we discovered through detective work that cryptlib wants the
    // signature's SID field to be the sha-1 hash of the SID.
    struct SignerInfo sigInfo;
    SignerInfo(&sigInfo, (ushort) 0);   /* init sigInfo */
    write_casn_num(&sigInfo.version.self, 3);
    // copy_casn(&sigInfo.version.self, &sigInfop->version.self); /* copy over 
    // */
    // copy_casn(&sigInfo.sid.self, &sigInfop->sid.self); /* copy over */
    write_casn(&sigInfo.sid.subjectKeyIdentifier, sid, sidsize);        /* sid 
                                                                         * hash */

    // copy over digest algorithm, signature algorithm, signature
    write_objid(&sigInfo.digestAlgorithm.algorithm, id_sha256);
    write_casn(&sigInfo.digestAlgorithm.parameters.sha256, (uchar *) "", 0);
    write_objid(&sigInfo.signatureAlgorithm.algorithm,
                id_rsadsi_rsaEncryption);
    write_casn(&sigInfo.signatureAlgorithm.parameters.rsadsi_rsaEncryption,
               (uchar *) "", 0);
    uchar *sig;
    int siglth = readvsize_casn(sigp, &sig);
    write_casn(&sigInfo.signature, &sig[1], siglth - 1);

    // now encode as asn1, and check the signature
    bsize = size_casn(&sigInfo.self);
    buf = (uchar *) calloc(1, bsize);
    encode_casn(&sigInfo.self, buf);
    ret = cryptCheckSignature(buf, bsize, pubkeyContext, hashContext);
    free(buf);

    // all done, clean up
    cryptDestroyContext(pubkeyContext);
    cryptDestroyContext(hashContext);
    cryptEnd();
    delete_casn(&rsapubkey.self);
    delete_casn(&sigInfo.self);

    // if the value returned from crypt above != 0, it's invalid
    return ret;
}

int main(
    int argc,
    char **argv)
{
    if (argc != 3)
        fatal(1, (char *)0);
    struct Certificate locert,
        hicert;
    struct CertificateRevocationList crl;
    Certificate(&locert, (ushort) 0);
    Certificate(&hicert, (ushort) 0);
    CertificateRevocationList(&crl, (ushort) 0);
    struct Blob blob;
    Blob(&blob, (ushort) 0);
    struct casn *tbsp,
       *sigp;
    struct AlgorithmIdentifier *algp;
    char *sfx = strchr(argv[1], (int)'.');
    int ansr;
    if (!strcmp(sfx, ".cer"))
    {
        tbsp = &locert.toBeSigned.self;
        algp = &locert.algorithm;
        sigp = &locert.signature;
        ansr = get_casn_file(&locert.self, argv[1], 0);
    }
    else if (!strcmp(sfx, ".crl"))
    {
        tbsp = &crl.toBeSigned.self;
        algp = &crl.algorithm;
        sigp = &crl.signature;
        ansr = get_casn_file(&crl.self, argv[1], 0);
    }
    else if (!strcmp(sfx, ".blb"))
    {
        tbsp = &blob.toBeSigned;
        algp = &blob.algorithm;
        sigp = &blob.signature;
        ansr = get_casn_file(&blob.self, argv[1], 0);
    }
    else
        fatal(4, argv[1]);
    if (ansr < 0)
        fatal(2, argv[1]);
    if (get_casn_file(&hicert.self, argv[2], 0) < 0)
        fatal(2, argv[2]);
    if (diff_objid(&algp->algorithm, id_sha_256WithRSAEncryption))
    {
        char oidbuf[80];
        read_objid(&algp->algorithm, oidbuf);
        fatal(5, oidbuf);
    }
    if (check_signature(tbsp, &hicert, sigp) < 0)
        fatal(0, "failed");
    fatal(0, "succeeded");
    return 0;
}
